# 2-0.変数と型
---

この章では、プログラミングの基本的な概念である「変数」と「型」について学ぶ。


### チュートリアル：Pythonでデータを表示してみよう

まずは手始めに、Pythonでデータを入力・表示してみよう。



In [1]:
print(5)

5


In [2]:
print(3.34)

3.34


In [3]:
print("AVILEN")

AVILEN


# 2-1. 変数の型
---

Pythonで扱うデータには、その性質を決めるさまざまな**「型」**が存在する。

大きくは以下の4つだ。

- 整数（int）　　例： `0`, `1`, `200`, `-10`
- 小数（float）　　例： `1.0`, `3.34`, `-0.05`, `1e4`
- 文字列（str）　　例: `"oya　sumi."`, `"a"`, `"ひらがな"`, `"30"`, `""`
- 真理値（bool）　　例: `True`, `False`

真理値（真偽値）とは，`True`(真)または`False`(偽)のどちらかのみの型で，主に条件分岐をするために使う。  

### 型の確認

変数の型を確認するには```type```関数を使う。

【例題】`False`の型を表示してみよう。

In [4]:
print(type(False))

<class 'bool'>


【問題】「kata」の型を表示してみよう。

- コードの中で文字列を書くときは'シングルクオーテーション'、もしくは"ダブルクオーテーション"で囲む必要がある

In [5]:
print(type("kata"))

<class 'str'>


【補足】プログラムでは実は`True`は数値の`1`, `False`は数値の`0`と等しい。

### 数値と文字列の違い

数値の場合、表記方法がint型、float型、str型の3通り存在する。

たとえば「`2`」と「`"2"`」はprintすると一緒の表記だが、コンピュータの中では違う型の変数として認識される。

pythonはこの型を自動で判別してくれるので便利である。

In [6]:
print(2)
print(2.0)
print("2")

2
2.0
2


In [7]:
#int
print(type(2))

#float
print(type(2.0))

#str
print(type("2"))

<class 'int'>
<class 'float'>
<class 'str'>


# 2-2.演算

---

次に、プログラミングに必要不可欠な「演算」の知識を身につけよう。

単純な計算機として演算を行ってみる。

### int型/float型の四則演算

【例題】Pythonで簡単な計算を行ってみよう。

In [8]:
print(2+3)
print(2.0+3.0)

5
5.0


In [9]:
print(2-3)
print(2.0-3.0)

-1
-1.0


In [10]:
print(2*3)
print(2.0*3.0)

6
6.0


In [11]:
print(2/3)
print(2.0/3.0)

0.6666666666666666
0.6666666666666666


【補足】int型の計算結果がfloat型(小数)になる場合、pythonは自動で判別してくれる(普通はできない。int型の計算結果はint型になる。)

In [12]:
print(type(2/3))

<class 'float'>


###  文字列(str)型の演算

【例題】次に、値が文字列のデータを足してみよう

In [14]:
print('今夜は月が'+'綺麗ですね')

今夜は月が綺麗ですね


- このように、文字列を足すと、2つの文字列を結合することができる

### 数値をstr型で表記した場合の演算

【例題】"2"と"3"を足してみよう

In [15]:
print('2'+'3')

23


数値もstr型で表記した場合は、計算ではなく結合の処理になることを覚えておこう。

【例題】"2"と"3"で四則演算ができるか試してみよう

In [17]:
print("2"-"3")

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [18]:
print("2"*"3")

TypeError: can't multiply sequence by non-int of type 'str'

In [19]:
print("2"/"3")

TypeError: unsupported operand type(s) for /: 'str' and 'str'

実行すると全てエラーになるはずだ。これはstr型の数値では四則演算ができないことを表している。

### int/float型とstr型の演算

次に、型の違うデータを足してみる。

【例題】"2"と3の足し算、および"2.0"と3.0の足し算を計算してみよう

In [20]:
print("2"+3)

TypeError: can only concatenate str (not "int") to str

In [24]:
print("2.0"+3.0)

TypeError: can only concatenate str (not "float") to str

型が違うデータ同士では、基本的に演算はできないことを覚えておこう。(int型とfloat型の計算は例外的に可能)

### 例外 : int型×str型をやってみると…？

実は例外的に、str型の掛け算(int型)はPythonだと計算できる。

実験として、「ピザ」という文字列を10回繰り返してみよう。

In [25]:
print("ピザ"*10)

ピザピザピザピザピザピザピザピザピザピザ


### その他の演算子

四則演算以外の計算もやってみよう。

#### 剰余(%)

```a%b```は、aをbで割った余りをint型で返す。

In [27]:
print(5%2)
print(type(5%2))

1
<class 'int'>


#### 累乗(**)

```a**b```は、aのb乗を表す。

In [28]:
print(3**2)
print(type(3**2))

9
<class 'int'>


### 計算の順序

計算の順序は基本的な数学の仕組みと同じで、掛け算・割り算が足し算・引き算より優先される。

また、括弧()を使うと優先順位を指定できる。

In [29]:
print(5+6*3)

23


In [30]:
print((5+6)*3)

33


# 2-3. 変数を使ってみよう

---

### 計算結果の使い回しや修正を行うと…

まず、変数を使わずに複雑な計算を行ってみよう。

In [31]:
print((3.1*10**3)**2)

9610000.0


この計算結果を他の計算に使いたい、というケースを考える。

そうすると、また上記の式を全て書かなければいけない。

In [32]:
print((3.1*10**3)**2 + 20000*1.25)

9635000.0


さらに、最初の計算結果を変更したいケースを考えよう。

In [33]:
print((2.1*10**3)**2)

4410000.0


こうすると、後ろのセルの計算式も全て書き換えないといけない。

In [34]:
print((2.1*10**3)**2+20000*1.25)

4435000.0


データサイエンスやAIの計算では、前のセルの計算結果を次のセルで使い回すことが多い。

そのときにいちいち前のセルの計算式を書いていくのは、非常に面倒くさいことがわかるだろう。

そこで役に立つのが「変数」という概念だ。

「変数」とは、データを保存する箱のような概念である。変数に何らかのデータを保存することを「格納」という。

### 数値データの格納

まずは最もよく使う数値の格納を学ぼう。

【例題】変数**```x```**に1を格納して、xの値を出力してみよう。

In [35]:
x=1
print(x)

1


もちろん整数だけでなく、小数も格納して出力できる。

【問題】 変数```param```に3.1415を格納して、変数paramを出力してみよう。

In [36]:
param=3.1415
print(param)

3.1415


### 変数の命名ルール

変数名には英大文字、英小文字、数字と"_"(アンダースコア)が使える。英大文字と小文字は区別されることに気をつけよう。

In [37]:
param=1
Param=-1

print(param)
print(Param)

1
-1


なお、変数名の最初が数字だとエラーが起きるので注意しよう。

In [38]:
2m=10

SyntaxError: invalid decimal literal (2164524361.py, line 1)

### 文字列の格納

変数には数値だけでなく、文字列も格納できる。

【例題】変数```asa```に「おはようございます！」と格納して出力してみよう。

In [40]:
asa='おはようございます！'
print(asa)

おはようございます！


【問題】変数```yoru```に「こんばんは。」と格納して出力してみよう。

In [41]:
yoru='こんばんは。'
print(yoru)

こんばんは。


### 変数の意義

そこで、例題の計算結果を変数```a```に格納してみよう。

In [46]:
a=(3.1*10**3)**2
print(a)

9610000.0


こうすることで、計算結果```a```を以降のセルの計算に利用することができる。

次に、この計算結果に```20000*1.25```を加えて変数```b```に格納してみよう。

In [47]:
b=a+20000*1.25
print(b)

9635000.0


さらに、```a```と```b```といった変数同士で計算することもできる。

2つを足し算して変数```c```に格納してみよう。

In [48]:
c=a+b
print(c)

19245000.0


どうだろう？先ほどと比べて、変数を使った式はとても楽に書けたのではないだろうか？

これが変数の効果で、記述量を大幅に減らしつつ、効率的に次のセルの計算ができるのだ。

【問題】変数```a```に2、変数```b```に3を格納して、```a+b```と```a-b```、```a*b```と```a/b```を計算しよう。

In [49]:
a=2
b=3

print(a+b)
print(a-b)
print(a*b)
print(a/b)

5
-1
6
0.6666666666666666


【問題】

アインシュタインの特殊相対性理論で、質量と等価なエネルギー量は$E = mc^2$で求められる。

りんご一個の重さを$m = 0.1 kg$、光速を$c = 2.9979×10^8 m/s$として、りんご一個の質量と等価なエネルギー量```E```を求めてみよう。

まずは必要な変数```m```と```c```を定義し、次にエネルギー量を計算して、変数```E```に格納・出力しよう。

In [51]:
m=0.1
c=2.9979*10**8

E=m*c**2
print(E)

8987404410000000.0


#  2-4. 変数の呼び出しと変数の上書き
---

【例題】 セルを跨いで```x```を「呼び出して」表示し、xの値を```"python"```に上書きしてみよう。

In [52]:
print(x)
x='python'
print(x)

1
python


- このように先ほどの【例題】で1が格納された変数`x`はセルを跨いで保存される
- カーネルの再起動で変数をリセットできる
- `=`は**「代入演算子」**の一つで、変数の値を「上書き保存」する

【問題】　xの値を"jupyter notebook"で上書きして、xを出力してみよう。

In [53]:
x='jupyter notebook'
print(x)

jupyter notebook


【問題】３つの文字列`a = "pen"`、 `b ="pineapple"`、　`c = apple`がある。次に`ab`を`a + b`とし、`ca`を`c + a`とする。最後に、`abca`を`ab + ca`として、`abca`出力したい。

以下のセルの抜けている部分を埋めてみよう

In [54]:
a='pen'
b='pineapple'
c='apple'

ab=a+b
ca=c+a
abca=ab+ca
print(abca)

penpineappleapplepen


#### 【補足】Pythonによる型の自動変換

実は他のプログラミング言語では、変数に対して型を指定する。

つまり、`x`という変数をstr型で指定した場合、その`x`の中身には文字列しか入れられない。

しかしPythonでは、型を宣言せずに変数を扱うため、同じ変数`x`の中身がstr型からint型に変わっても対応できる。

このとき、自動で型を変換してくれる。

In [55]:
x='jupyter notebook'
print(x)
print(type(x))
x=10
print(x)
print(type(x))

jupyter notebook
<class 'str'>
10
<class 'int'>


## 2-5. bool型と比較演算子
【例題】比較演算子を使うと，指定した条件が正しいのか，間違っているのかを判別できる．

In [56]:
x=2
y=1
print(x>1)

True


- 比較演算子の結果はbool型となり、`True`は正しい(真)、`False`は間違っている(偽)を表している
- 比較演算子は他に以下のものがある。  
    - 「==」左右が等しいかどうか  
    - 「!=」左右が異なっているかどうか  
    - 「<=」左の値が右の値以下かどうか  

【例題】```a=1==1```および```b=1==2```を計算してみよう

In [57]:
a=1==1
print(a)

True


これは```a```という変数に```1```と```1```というint型の比較結果が入る。左右が等しいので```True```が返されている。

In [58]:
b=1==2
print(b)

False


これは```b```という変数に```1```と```2```の比較結果が入り、左右が等しくないので```False```が返される。

なお、下記のように書くとエラーになる。

In [60]:
a=1=2

SyntaxError: cannot assign to literal (3621963478.py, line 1)

【問題】 `x`に`0.3`を代入し、`y`に`0.1*3`を代入して、`x`と`y`が等しいか比較演算子を使って判定しよう．  
その結果を変数`z`に格納して，`z`を表示してみよう．

In [74]:
x=0.3
y=0.1*3
print('x:',x)
print('y:',y)
print('x==y:',x==y)
z=x==y
print('z:',z)

x: 0.3
y: 0.30000000000000004
x==y: False
z: False


- 比較演算子の結果はbool型なので変数に格納できる
- 実はコンピュータでは、2進数と10進数の変換の関係で、小数の計算は厳密な精度ではできない。
- 例えば`0.1`の3倍は`0.30000000000000004`になってしまう.

### 〜演算子一覧〜

- 比較演算子（`==`, `!=`, `<`, `>`, `<=`, `>=`, `is`, `<>`）<br>
- 代入演算子(`=`, `+=`, `-=`, `*=`, `/=`, `%=`, `**=`, `//=`, `&=`, `|=`, `^=`, `<<=`, `>>=`)<br>
- ビット演算子(`~`, `&`, `|`, `^`, `<<`, `>>`)<br>
- ブール演算子(`and`, `or`, `not`)<br>

どんな演算ができるか気になる場合は一度試してみよう。（もちろん暗記する必要はない）